### DISCUSSION TITLE
Dropdown list options separator, text when set from MQTT

### DISCUSSION THREAD

**Discussion by nagyrobi** [HASwitchPlate/openHASP#105] [Q&A] [answered]
According to the [docs](https://fvanroie.github.io/hasp-docs/#objects/#dropdown-list) `\n` has to be used between option items of a dropdown list. 
I tried sending through MQTT to my `hasp/plate/command/p2b4.options` the string `a\nb\nc\n\d` and instead of dropping them to different items, it created a single item with \n's in it:
![kép](https://user-images.githubusercontent.com/1550668/112147179-b0c99900-8bdc-11eb-9d5d-491f090a2a65.png)
Another problem that the object does not update on the screen after posting the MQTT command, you have to switch page back and forth to see the changes.
if set from JSONL, it's OK.

**Comment by nagyrobi** [comment]
Similar issue:
setting "text":"\uF185" should display a sun icon, instead, it displays `]uF185` when set through MQTT.
if set from JSONL, it's OK.
![kép](https://user-images.githubusercontent.com/1550668/112302239-2431e000-8c9b-11eb-9596-5f65d0dfb67f.png)
left button set from MQTT, right button set from uploaded .jsonl file:
`{"obj":"btn","id":2,"x":10,"y":40,"w":105,"h":90,"toggle":true,"text":"\uF185","text_font":28,"mode":"break","align":1}`

**Reply by arovak** [reply]
I cannot reproduce with the latest build of the master branch. Can you show the logs from the device as when it received the MQTT command?

![afbeelding](https://user-images.githubusercontent.com/56759291/112306433-1599f780-8ca0-11eb-986d-137404c0caad.png)

`{"page":1,"obj":"btn","id":2,"x":10,"y":40,"w":105,"h":90,"toggle":"true","text":"\uF185","text_font":28,"mode":"break","align":1}
{"page":1,"obj":"dropdown","id":3,"x":10,"y":150,"w":150,"h":40,"options":"Line 1\nLine 2\nLine 3\nline 4"}
`

log file of my test device
`[Mar 24 12:56:40.273][81008/155944 48][54500/54680  1] MQTT RCV: hasp/test_plate/command/jsonl = {"page":1,"obj":"btn","id":2,"x":10,"y":40,"w":105,"h":90,"toggle":"true","text":"\uF185","text_font":28,"mode":"break","align":1}
{"page":1,"obj":"dropdown","id":3,"x":10,"y":150,"w":150,"h":40,"options":"Line 1\nLine 2\nLine 3\nline 4"}`

**Reply by nagyrobi** [reply]
You're using `jsonl` command. Try use `text` or `options` command directly, not `jsonl`.
Like I described in my first post.

**Reply by fvanroie** [reply]
> To use an icon in a text you need to prefix the UTF-8 value with \u. To ensure proper decoding the payload should be used with a json or jsonl command.

As described [here](https://fvanroie.github.io/hasp-docs/#fonts/#custom-fonts).

`json` and `jsonl` commands decode the escape characters into the correct UTF-8 bytes.

If you use `.text` or `.options` directly, you need to make sure that the bytes are already encoded properly in the mqtt payload.

**Reply by nagyrobi** [reply]
\u escape character is not recognized unfortunately in text, options. Home Assistant templates don't offer more than that imho.

**Reply by arovak** [reply]
> 
> 
> You're using `jsonl` command. Try use `text` or `options` command directly, not `jsonl`.
> Like I described in my first post.

still cannot reproduce with json.
`[Mar 24 16:03:20.874][81008/155960 48][53948/54196  1] MQTT RCV: hasp/test_plate/command/json = ['p1b2.text=\uF185', '
p1b3.options=test 1\ntest 2\ntest 3']
[Mar 24 16:03:20.883][81008/155864 48][53948/54196  1] MSGR: p1b2.text=…
[Mar 24 16:03:20.896][81008/155864 48][53948/54196  1] MSGR: p1b3.options=test 1
                                                                                test 2
                                                                                      test 3
`
![afbeelding](https://user-images.githubusercontent.com/56759291/112333316-c7462200-8cba-11eb-9db3-7c65e839e00e.png)

are you sure HA is sending the message in the correct form? What does the log of the device show?

**Comment by fvanroie** [comment]
Any modern programming language has the ability to encode/decode unicode bytes quite easily.

There are 2 distinct paths that the payload can follow: One is inside a json *wrapper* and the other one is raw bytes.
Forcing the raw bytes to also pass through a json *'parser and decoder'* opens up other issues:
- hasp-lvgl operates on the raw mqtt payload as much as possible for optimizing *speed* and *memory* resources.
- An unnecessary json conversion just adds *overhead* if there are no escape characters needed.
- By adding escape sequences, you also need to use an unescape sequence for any backslashes you have.

I just don't think adding json parsers is the correct solution here, since this can be solved in other ways that don't burden the MCU.

**Reply by nagyrobi** [reply]
Understood, thanks. Especially that it [turned out](https://github.com/dgomes/hasp-lvgl-custom-component/issues/20) that Home Assistant can do this. 

The biggest problem of all these things is lack of proper documentation. I've been using and contributing to Home Assistant for 2 years now and I've never came across this. They didn't document this @dgomes must be a super-expert above everybody in their community.

**Reply by fvanroie** [reply]
Yes, he's doing a fabulous job with the custom component. We're all still learning here.
The take-away is that we can document this ourselves and help make a better integration possible!

**Reply by dgomes** [reply]
I just google "escape string jinja2" 😅

**Reply by htvekov** [reply]
Yep, Google is your friend (when they not selling your personal data 🤣)

It also took me some time to get this template working in HA. Never had to escape an escape character before. But HA's yaml parser 'eats' the first one processing the line.

```yaml
payload_template: "{% if trigger.to_state.state == 'paused' %}\n  {{ '\\uf04b'\
        \ }}\n{% elif trigger.to_state.state == 'playing' %}\n  {{ '\\uf04c' }}\n\
        {% else %}\n  {{ '\\uf04d' }}\n{% endif %}" 
```

**Reply by fvanroie** [reply]
> But HA's yaml parser 'eats' the first one processing the line.

I think it's because you use `"` as outer quotes. Maybe try to use `'` as outer quotes and `"` for inner quotes.
That would explain the behavior.

**Reply by htvekov** [reply]
Spot on with that observation, @fvanroie.
Just tested, and indeed no  need to 'escape' the escape character reverting to` ' `as outer quotes. YAML syntax is a bitch sometimes (most times)... 😉 There also the pipe possibility as you mention.

**Comment by fvanroie** [comment, marked as answer]
Besides appending `|e` *(pipe symbol + `e`)* at the end of the template, the escape character can be either `\x`, `\u` or `\U`:

- Code points up to `0xFF` should be encoded as `"\xE6"` in the template.

- Code points up to `0xFFFF` should be encoded as `"\uE6E8"` in the template.
  __Note:__ Use lowercase `\u` and exactly 4 hexadecimal digits.

- Code points above `0xFFFF` *must* be encoded as `"\U0001F5E9"` in the template.
  __Note:__ Use capital `\U` and exactly 8 hexadecimal digits.
